feat(linter): add `oxc/no-map-spread`

DonIsaac opened the pull request:
Adds a new performance rule that disallows using object and array spreads in `map` and `flatMap` to modify/merge arrays and objects.
I've currently added it to `nursery` so we can battle-harden it before making it generally available as a `perf` rule.

```ts
// needless O(n) spread on `x`, making this an O(n^2) operation
const x = arr.map(x => {
  return { ...x, foo: getFoo(x.id) }
})

// Object.assign is better here
const x = arr.map(x => {
  return Object.assign(x, { foo: getFoo(x.id) })
})
```

DonIsaac commented:
* **#6751** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6751?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

camc314 reviewed (commented):
this looks great.

camc314 commented on crates/oxc_linter/src/rules/oxc/no_map_spread.rs:
```diff
@@ -0,0 +1,545 @@
+            // Spread does not occur within `map`
+            "let extras = { ...foo, ...bar }; let a = b.map(x => extras)",
+            None,
+        ),
```
i would suggest adding
```ts
function foo(arr) {
    return arr.map(x => ({ ...x }))
}
```

i don't know whether this would be reported currently, but it shuold not be as we don't know whether the fn arg could be used at the calcite

camc314 replied:
sounds good to me

DonIsaac replied:
I've added `ignoreArgs`. It defaults to `true`.

camc314 commented on crates/oxc_linter/src/rules/oxc/no_map_spread.rs:
```diff
@@ -0,0 +1,545 @@
+        return None;
+    }
+
+    let arg = call_expr.arguments.first()?.as_expression()?;
```
use without_parenthes/get_inner_expr ?

camc314 commented on crates/oxc_linter/src/rules/oxc/no_map_spread.rs:
```diff
@@ -0,0 +1,545 @@
+        let Some(mapper) = get_map_callback(call_expr) else {
+            return;
+        };
+        // let Some((is_expression, body)) = get_map_callback(call_expr) else {
+        //     return;
+        // };
+
```
```suggestion
```

camc314 reviewed (approved):
this looks good to me - there's some dead code that could be removed.

i also wonder whether it's worth either:
a. `SpreadInReturnVisitor::<'a, '_>::iter_spreads` converting this to follow the same pat as no_accumulating_spread to fine spreads
b. converting no_accumulating spread to use `SpreadInReturnVisitro`

camchenry commented on crates/oxc_linter/src/rules/oxc/no_map_spread.rs:
```diff
@@ -0,0 +1,578 @@
+
+fn no_map_spread_diagnostic(map_call: Span, spread: &Spread<'_, '_>) -> OxcDiagnostic {
+    let spans = spread.spread_spans();
+    assert!(!spans.is_empty());
+    let mut spread_labels = spread.spread_spans().into_iter(); // .map(LabeledSpan::from);
+    let first_message = if spans.len() == 0 {
```
This seems contradictory: we are asserting `spans` is not empty, but then checking if spans length is zero? Do we need to assert here? (If so, could we use `debug_assert!`?)

DonIsaac replied:
That should be 1, good catch.

camchenry commented on crates/oxc_linter/src/rules/oxc/no_map_spread.rs:
```diff
@@ -0,0 +1,578 @@
+        // we don't care if the symbol is overwritten
+        .filter(|r| r.is_read())
+        // Find where the symbol was declared
+        .map(|r| ctx.semantic().nodes().get_node(r.node_id())) //
```
```suggestion
        .map(|r| ctx.semantic().nodes().get_node(r.node_id()))
```

camchenry commented on crates/oxc_linter/src/rules/oxc/no_map_spread.rs:
```diff
@@ -0,0 +1,578 @@
+            // Spread does not occur within `map`
+            "let extras = { ...foo, ...bar }; let a = b.map(x => extras)",
+            None,
+        ),
```
Would probably be a good idea to add some "close-but-not-quite" cases like if `map` is a local function call (I think we already handle this, but always good to make sure)

```
let x = {}
function map(f) {}
map(x => ({ ...x }))
```

or map just returns an object:

```
a.map(x => ({ foo: x }))
```

DonIsaac replied:
Added cases for custom map definitions

camchenry commented on crates/oxc_linter/src/snapshots/no_map_spread.snap:
```diff
@@ -0,0 +1,144 @@
+
+  × Expected `,` but found `}`
+   ╭─[no_map_spread.tsx:1:49]
+ 1 │ let a = b.map(({ x, y }) => ({ ...(cond ? x : y })
+   ·                                                 ┬
+   ·                                                 ╰── `,` expected
```
Looks like this is missing a paren

mysteryven reviewed (commented):
LGTM

DonIsaac commented:
### Merge activity

* **Oct 26, 1:40 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

DonIsaac committed:
feat(linter): add `oxc/no-map-spread` (#6751)

Adds a new performance rule that disallows using object and array spreads in `map` and `flatMap` to modify/merge arrays and objects.
I've currently added it to `nursery` so we can battle-harden it before making it generally available as a `perf` rule.

```ts
// needless O(n) spread on `x`, making this an O(n^2) operation
const x = arr.map(x => {
  return { ...x, foo: getFoo(x.id) }
})

// Object.assign is better here
const x = arr.map(x => {
  return Object.assign(x, { foo: getFoo(x.id) })
})
```
